### PR TITLE
Use weak handle for the back reference to EventProvider

### LINKED ISSUE #80450: EventListener unloadability
#79970 has broken unloadability for a couple of coreclr tests on both Windows and Unix. The unloading is blocked due to root like this (the root path differs a bit test by test):
```
(lldb) gcroot -all 00007fbf6a813690
HandleTable:
    00007FFFF4AC1390 (strong handle)
    -> 00007FBF6A81A800 System.Diagnostics.Tracing.EventPipeEventProvider
    -> 00007FBF6A81A828 System.Diagnostics.Tracing.EventEnableCallback
    -> 00007FBF6A81A7A8 System.Diagnostics.Tracing.EventSource+OverrideEventProvider
    -> 00007FBF6A816538 System.Threading.Tasks.ParallelEtwProvider
    -> 00007FBF6A813690 System.Reflection.LoaderAllocator

Found 1 roots.
```

The failing tests are:
JIT/Performance/CodeQuality/BenchmarksGame/k-nucleotide/k-nucleotide-9
JIT/Performance/CodeQuality/BenchmarksGame/spectralnorm/spectralnorm-3
JIT/Performance/CodeQuality/BenchmarksGame/mandelbrot/mandelbrot-7
JIT/Performance/CodeQuality/Roslyn/CscBench
JIT/Regression/JitBlue/GitHub_19361/GitHub_19361
JIT/superpmi/superpmicollect/CscBench
profiler/multiple/multiple
Regressions/coreclr/GitHub_45929/test45929
tracing/eventlistener/eventlistener

### PR DESCRIPTION
**Pull request by jkotas**
* Use weak handle for the back reference to EventProvider. This avoids leak when the event provider is orphaned without explicit unregistration.
 
 * Delete unnecessary IEventProvider interface. Repurpose no-op event provider as a base class

Fixes #80450

### CODE REVIEW

**Comment by jkotas** [comment]
@janvorli Is it possible to trigger the unloadability test run from CI?

**Comment by janvorli** [comment]
@jkotas I am not sure if that's possible. But we have weekly runs that execute those (https://dev.azure.com/dnceng-public/public/_build?definitionId=121)

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs [hunk 1]
```diff
@@ -5,27 +5,29 @@
 
 namespace System.Diagnostics.Tracing
 {
-    internal sealed class EventPipeEventProvider : IEventProvider
+    internal sealed class EventPipeEventProvider : EventProviderImpl
     {
-        private EventEnableCallback? _enableCallback;
+        private WeakReference<EventProvider> _eventProvider;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventPipeEventProvider.cs:10]: Nit: readonly (here and in the other similar fields)?

**Comment by jkotas** [comment]
I have manually triggered a run - the tests that you have flagged are passing.

**Review by noahfalk — APPROVED** [review]
LGTM, thanks Jan!

@davmason - you might want to take a look as well.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**